core/src/main/java/io/grpc/internal/ServerCallImpl.java

```diff
@@ -73,6 +104,33 @@ public void request(int numMessages) {
   public void sendHeaders(Metadata headers) {
     checkState(!sendHeadersCalled, "sendHeaders has already been called");
     checkState(!closeCalled, "call is closed");
+
+    headers.removeAll(MESSAGE_ENCODING_KEY);
+    if (compressor == null) {
+      compressor = Codec.Identity.NONE;
+      if (inboundHeaders.containsKey(MESSAGE_ACCEPT_ENCODING_KEY)) {
+        String acceptEncodings = inboundHeaders.get(MESSAGE_ACCEPT_ENCODING_KEY);
+        for (String acceptEncoding : ACCEPT_ENCODING_SPLITER.split(acceptEncodings)) {
+          Compressor c = compressorRegistry.lookupCompressor(acceptEncoding);
+          if (c != null) {
+            compressor = c;
+            break;
+          }
+        }
+      }
+    }
+    inboundHeaders = null;
```

Comment:
Null it out in the case of close() as well?

Technically, neither sendHeaders() nor close() is required to be called, because the client could cancel the call.